WordPress/Classes/Stores/StatsInsightsStore.swift

```diff
@@ -981,3 +985,66 @@ private extension InsightStoreState {
     }
 
 }
+
+// MARK: - iOS 14 Widgets Data
+private extension InsightStoreState {
+
+    private func saveHomeWidgetTodayData(stats: TodayWidgetStats) {
+
+        guard let siteID = SiteStatsInformation.sharedInstance.siteID else {
+            return
+        }
+
+        var homeWidgetTodayCache = HomeWidgetTodayData.read() ?? initializeHomeWidgetTodayData()
+
+        guard let oldData = homeWidgetTodayCache[siteID.intValue] else {
+            DDLogError("HomeWidgetToday: Failed to find a matching site")
+            return
+        }
+        // TODO - TODAYWIDGET: it might be better to move the blog updates in SiteStatsInformation
+        let blogService = BlogService(managedObjectContext: ContextManager.shared.mainContext)
+
+        guard let blog = blogService.blog(byBlogId: siteID) else {
+            DDLogError("HomeWidgetToday: the site does not exist anymore")
+            // if for any reason that site does not exist anymore, remove it from the cache.
+            homeWidgetTodayCache.removeValue(forKey: siteID.intValue)
+            HomeWidgetTodayData.write(data: homeWidgetTodayCache)
+            return
+        }
+        // refresh stats and update any blog info, if they had changed
+        homeWidgetTodayCache[siteID.intValue] = HomeWidgetTodayData(siteID: siteID.intValue,
+                                                           siteName: blog.title ?? oldData.siteName,
+                                                           url: blog.url ?? oldData.url,
+                                                           timeZoneName: blogService.timeZone(for: blog).identifier,
+                                                           date: Date(),
+                                                           stats: stats)
+
+        HomeWidgetTodayData.write(data: homeWidgetTodayCache)
+        if #available(iOS 14.0, *) {
+            WidgetCenter.shared.reloadTimelines(ofKind: "WordPressHomeWidgetToday")
+        }
+    }
+
+    // TODO - TODAYWIDGET: this method will likely be exposed in the future to handle the creation of the file at login
+    private func initializeHomeWidgetTodayData() -> [Int: HomeWidgetTodayData] {
+
+        let blogService = BlogService(managedObjectContext: ContextManager.shared.mainContext)
+
+        return blogService.blogsForAllAccounts().reduce(into: [Int: HomeWidgetTodayData]()) {
+            if let blogID = $1.dotComID,
+               let url = $1.url,
+               let blog = blogService.blog(byBlogId: blogID) {
+
+                let title = ($1.title ?? url).isEmpty ? url : $1.title ?? url
```

Comment:
Can this just be `$1.title ?? url`? 🤔

Reply:
+1

Reply:
If I did that, we would not account the case of a title that is not nil, but empty. What we want, instead, is that if the site has an empty title, we display the url as a title

Reply:
While I now can see why it's done this way, I have to admit it wasn't immediately obvious by reading it.

Any thoughts on rewriting this to make it a bit more readable and matinainable, even if it means slightly less compact?

As an example:

```swift
let title = (element.title != nil && ! element.title.isEmpty) ? element.title : url
```

Reply:
This is more readable, but does not completely work, as it returns a `String?` instead of a `String`

Reply:
I added a comment to this line to describe what the code does, hoping that it helps to better understand what it does